### PR TITLE
Replace Mockito spy usage in QueryIndexingTest [HZ-2491]

### PR DESCRIPTION
**Pull request by JamesHazelcast**
We have recently been encountering a variety of edge-case issues with Mockito related failures in some of our tests - these tests are generally non-deterministic, but surface during long-running test jobs on Jenkins. There has been extensive investigation of these issues by multiple team members, but the issues always point towards an issue within Mockito itself (although exact causes have not been identified).

The QueryIndexingTest encounters the same failures consistently while running on a Sonar code coverage Jenkins job. The failures are always related to instantiating a _serialized spy_ instance, with various debugging sessions showing that non-serialized spy instances are created successfully. As with other similar Mockito failures, this failure is reproduced consistently in long-running tests, but not in isolation.

Due to the functionality of this being simple in this test, we have decided to avoid spending more time investigating the specifics of why Mockito is struggling here and simply replace the problematic Mockito functionality with our own implementation. In this test that is a simple wrapper for the `Employee` class which uses a `Map` keyed by method names to an `AtomicInteger` which counts the invocations of each method.

Fixes https://github.com/hazelcast/hazelcast/issues/23645

### CODE REVIEW

**Review by TomaszGaweda — APPROVED** [review]
Nice finding :)

File: hazelcast/src/test/java/com/hazelcast/query/SampleTestObjects.java [hunk 1]
```diff
@@ -391,6 +394,86 @@ public String toString() {
...
+        }
+
+        public SpiedEmployee(long id, String name, String city, int age, boolean live, double salary) {
+            super(id, name, city, age, live, salary);
+        }
+
+        public SpiedEmployee() {
+        }
+
+        public long getId() {
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/query/SampleTestObjects.java:426]: Question: if two threads will call getId, there's a chance that the second change will override invocationCounters("getId"). Is it ever called from multiple threads? If no, then better to change whole map to HashMap

Comment (JamesHazelcast) [reply hazelcast/src/test/java/com/hazelcast/query/SampleTestObjects.java:426]: I wrote the `SpiedEmployee` class to be fairly dynamic and usable in multiple tests, should we need it in the future - with that in mind, I opted to use a `ConcurrentHashMap` with `AtomicInteger` values to keep it thread-safe and ensure we don't encounter any race conditions or data loss when using this implementation (even in a multi-threaded environment). Due to this class only being used during test runs, the performance overhead from this should not be an issue.

**Review by ahmetmircik — APPROVED** [review]
